ports/stm32f4/common-hal/busio/I2C.c

```diff
@@ -35,51 +35,46 @@
 #include "supervisor/shared/translate.h"
 #include "common-hal/microcontroller/Pin.h"
 
-STATIC bool reserved_i2c[3];
-STATIC bool never_reset[3];
+//arrays use 0 based numbering: I2C1 is stored at index 0
+#define MAX_I2C 3
+STATIC bool reserved_i2c[MAX_I2C];
+STATIC bool never_reset_i2c[MAX_I2C];
 
-void i2c_reset(void) {
-    //Note: I2Cs are also forcibly reset in construct, due to silicon error
-    #ifdef I2C1
-        reserved_i2c[0] = false;
-        __HAL_RCC_I2C1_CLK_DISABLE(); 
-    #endif
-    #ifdef I2C2
-        reserved_i2c[1] = false;
-        __HAL_RCC_I2C2_CLK_DISABLE(); 
-    #endif
-    #ifdef I2C3
-        reserved_i2c[2] = false;
-        __HAL_RCC_I2C3_CLK_DISABLE(); 
-    #endif
-}
-
-void common_hal_busio_i2c_never_reset(busio_i2c_obj_t *self) {
-    for (size_t i = 0 ; i < MP_ARRAY_SIZE(mcu_i2c_banks); i++) {
-        if (self->handle.Instance == mcu_i2c_banks[i]) {
-            never_reset[i] = true;
+#define ALL_CLOCKS 0xFF
+STATIC void i2c_clock_enable(uint8_t mask);
+STATIC void i2c_clock_disable(uint8_t mask);
 
-            never_reset_pin_number(self->scl->pin->port, self->scl->pin->number);
-            never_reset_pin_number(self->sda->pin->port, self->scl->pin->number);
-            break;
+void i2c_reset(void) {
+    uint16_t never_reset_mask = 0x00;
+    for(int i = 0; i < MAX_I2C; i++) {
+        if (!never_reset_i2c[i]) {
+            reserved_i2c[i] = false;
+        } else {
+            never_reset_mask |= 1 << i;
         }
     }
+    i2c_clock_disable(ALL_CLOCKS & ~(never_reset_mask));
 }
 
-
 void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
         const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout) {
 
     //match pins to I2C objects
     I2C_TypeDef * I2Cx;
+    uint8_t sda_len = sizeof(mcu_i2c_sda_list) / sizeof(*mcu_i2c_sda_list);
```

Comment:
While it's not used consistently, it's worth noting that there's a macro for this...
py/misc.h:#define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))

Reply:
noice I'll use it.